File: frontend/src/api/forge.ts
import apiClient from './client'

// Transform workbench ("Forge"): a CyberChef-style recipe engine. The user
// pastes a value, stacks an ordered list of operations, and the backend runs
// them left-to-right. Mirrors backend/internal/api/handlers transform contract.

export type ForgeArgType = 'string' | 'int' | 'bool' | 'select'

export interface ForgeArg {
  key: string
  label: string
  type: ForgeArgType
  default?: string
  options?: string[]
  help?: string
}

export interface ForgeOp {
  name: string
  category: string
  description: string
  args?: ForgeArg[]
}

export interface ForgeRecipeStep {
  op: string
  args?: Record<string, string>
}

export interface ForgeStep {
  op: string
  output: string
  binary: boolean
  bytes: number
  error?: string
}

export interface ForgeResult {
  steps: ForgeStep[]
  output: string
  binary: boolean
  bytes: number
}

interface ApiResponse<T> {
  success: boolean
  error?: string
  data: T
}

export interface ForgeRunResult {
  success: boolean
  error?: string
  result: ForgeResult
}

export const forgeApi = {
  // The catalogue of available operations, grouped client-side by category.
  operations: async (): Promise<ForgeOp[]> => {
    const { data } = await apiClient.get<ApiResponse<ForgeOp[]>>('/tools/transform/operations')
    return data.data
  },

  // Run a recipe. A failing step still returns HTTP 200 with success:false, an
  // error, and a partial trace (steps up to the failure) — surface both.
  run: async (input: string, recipe: ForgeRecipeStep[]): Promise<ForgeRunResult> => {
    const { data } = await apiClient.post<ApiResponse<ForgeResult>>('/tools/transform', {
      input,
      recipe,
    })
    return { success: data.success, error: data.error, result: data.data }
  },
}

File: backend/internal/api/handlers/forge.go
package handlers

import (
	"net/http"

	"github.com/analysishub/backend/internal/forge"
	"github.com/gin-gonic/gin"
)

// ListForgeOperations returns the transform palette: every available operation,
// its category, description and argument spec, so the UI can build the recipe
// builder without hard-coding the list.
//
// GET /api/v1/tools/transform/operations
func ListForgeOperations(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"success": true, "data": forge.Operations()})
}

// RunForgeRecipe applies an ordered recipe of operations to an input and returns
// the per-step trace plus the final output. Pure compute, no DB — the CyberChef-
// style decode/encode/crypto workbench.
//
// POST /api/v1/tools/transform
// Body: { "input": "...", "recipe": [ { "op": "From Base64", "args": {...} }, ... ] }
func RunForgeRecipe(c *gin.Context) {
	var req struct {
		Input  string             `json:"input"`
		Recipe []forge.RecipeStep `json:"recipe"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "error": err.Error()})
		return
	}

	res, err := forge.Run(req.Input, req.Recipe)
	if err != nil {
		// A step failure is a normal user outcome (wrong key, bad input), so the
		// partial trace is still returned alongside the error for diagnosis.
		c.JSON(http.StatusOK, gin.H{"success": false, "error": err.Error(), "data": res})
		return
	}
	c.JSON(http.StatusOK, gin.H{"success": true, "data": res})
}
